src/transaction/mod.rs

```diff
@@ -53,52 +86,91 @@ pub enum Mutation {
+        let key = key.into();
+        match self.get_from_mutations(&key) {
+            MutationValue::Determined(value) => Ok(value),
+            MutationValue::Undetermined => self.snapshot.get(key).await,
+        }
     }
 
-    /// Gets the values associated with the given keys.
+    /// Gets the values associated with the given keys. The returned iterator is in the same order
+    /// as the given keys.
     ///
     /// ```rust,no_run
     /// # #![feature(async_await)]
-    /// # use tikv_client::{KvPair, Config, TransactionClient};
+    /// # use tikv_client::{Key, Value, Config, transaction::Client};
     /// # use futures::prelude::*;
+    /// # use std::collections::HashMap;
     /// # futures::executor::block_on(async {
-    /// # let connecting_client = TransactionClient::connect(Config::new(vec!["192.168.0.100", "192.168.0.101"]));
+    /// # let connecting_client = Client::connect(Config::new(vec!["192.168.0.100", "192.168.0.101"]));
     /// # let connected_client = connecting_client.await.unwrap();
     /// let mut txn = connected_client.begin().await.unwrap();
     /// let keys = vec!["TiKV".to_owned(), "TiDB".to_owned()];
-    /// let req = txn.batch_get(keys);
-    /// let result: Vec<KvPair> = req.await.unwrap();
+    /// let result: HashMap<Key, Value> = txn
+    ///     .batch_get(keys)
+    ///     .await
+    ///     .unwrap()
+    ///     .filter_map(|(k, v)| v.map(move |v| (k, v))).collect();
     /// // Finish the transaction...
     /// txn.commit().await.unwrap();
     /// # });
     /// ```
     pub async fn batch_get(
         &self,
-        _keys: impl IntoIterator<Item = impl Into<Key>>,
-    ) -> Result<Vec<KvPair>> {
-        unimplemented!()
+        keys: impl IntoIterator<Item = impl Into<Key>>,
+    ) -> Result<impl Iterator<Item = (Key, Option<Value>)>> {
+        let mut keys_from_snapshot = Vec::new();
+        let mut results_in_buffer = keys
+            .into_iter()
+            .map(|key| {
+                let key = key.into();
+                let mutation_value = self.get_from_mutations(&key);
+                if let MutationValue::Undetermined = mutation_value {
+                    keys_from_snapshot.push(key.clone());
+                }
+                (key, mutation_value)
+            })
+            .collect::<Vec<_>>()
+            .into_iter();
+        let mut results_from_snapshot = self
+            .snapshot
+            .batch_get(keys_from_snapshot)
+            .await?
+            .peekable();
+        Ok(std::iter::from_fn(move || {
+            let (key, mutation_value) = results_in_buffer.next()?;
+            match mutation_value {
+                MutationValue::Determined(value) => Some((key, value)),
+                MutationValue::Undetermined => match results_from_snapshot.peek() {
+                    Some((key_from_snapshot, _)) if &key == key_from_snapshot => {
+                        results_from_snapshot.next()
+                    }
```

Comment:
I think it would be good to add a comment inside this method somewhere to explain what you're doing.

Reply:
Sorry, I find I am doing wrong here. These logic should belong to the batch_get in snapshot instead of here.